### ISSUE TITLE
Logical `and` is not interpreted when figuring out remaining expression types

### ISSUE THREAD

**Issue by gh-andre** [python/mypy#15332]
**Bug Report**

When a logical `and` is used to avoid introducing `None` into the code that follows, parts of the conditional expression are not evaluated together to figure out the impossible combination of types, so this error is reported:

> error: Incompatible types (expression has type "Optional[str]", TypedDict item "x" has type "str")  [typeddict-item]

**To Reproduce**

This code reports the error above.
```python
from typing_extensions import TypedDict

class X(TypedDict):
    x: str
    
def get_v(n: str) -> str|None:
    return "ABC"
    
v: str|None = get_v("X")

if (v2 := get_v("Y")) is None and v is None:
    raise Exception("Both v and v2 cannot be None")
    
if v2 is None:
    v2 = v
    
d: X = {
    "x": v2            # <<< error
}
```

**Expected Behavior**

`v2` is guaranteed not to be `None` when the `x` field is being assigned because `v2` and `v` cannot be both `None`.

**Actual Behavior**

`v2` is reported as if it can be `None` when `x` is being assigned.

**Your Environment**

- Mypy version used: 1.3.0
- Mypy command-line flags: --strict 
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.10

The error disappears if `and` is replaced with a nested conditional statement, like this:
```python
if (v2 := get_v("Y")) is None:
    if v is None:
        raise Exception("Both v and v2 cannot be None")
    v2 = v
```

**Comment by ikonst** [comment]
Mypy tracks expressions' types when they're _unconditionally_ narrowed. It doesn't keep track of conditions like "v2 is optional right now, but if any subsequent code narrows v, then narrow v2".

I don't think there's anything in the mypy docs clarifying this.

**Comment by ikonst** [comment]
Is this basically pyright's [Aliased Conditional Expressions](https://github.com/microsoft/pyright/blob/main/docs/type-concepts-advanced.md#aliased-conditional-expression)?

**Comment by erictraut** [comment]
@ikonst, no, this isn't an example of an aliased conditional expression. This is a case where the user expects a type checker to understand "conditional narrowing" where the type of one variable (`v2`) is conditioned on the type of a second variable (`v`). This is similar to the concept of "entanglement" in quantum physics. Tracking these sort of cross-variable conditions in a type checker would add significant complexity and performance overhead and would be computationally infeasible in all but the most basic cases. I'm not aware of any type checkers or compilers that attempt to do this.

In my experience, there's almost always a straightforward way to restructure the code to eliminate the need for "entanglement" between variables. IMO, it's reasonable to ask developers to make small accommodations in their code to work with a type checker. In this particular example, the OP identified a straightforward workaround.

**Comment by gh-andre** [comment]
Well, FWIW, TypeScript will report a similar error:
```typescript
type X = {
    x: string;
}
    
function get_v(n: string) : string|null
{
    return "ABC";
}
    
let v: string|null = get_v("X");
let v2: string|null

if ((v2 = get_v("Y")) == null && v == null)
    throw Error("Both v and v2 cannot be None")
    
if(v2 == null)
    v2 = v
    
let d: X = {
    x: v2           // TS2322: Type 'string | null' is not assignable to type 'string'
}
```

@erictraut Workarounds for this may result in code that is uglier than in the example. Consider a global configuration with a bunch of scoped configurations optionally overriding global values, so values may come from either location and both being `None` is flagged as invalid when the configuration is loaded. Using a conditional expression to get the correct value where needed makes such code very readable, while having multiline statements or obscure methods with overrides makes it less maintainable.

Having said that, I would agree that complexity to implement this probably isn't worth the benefit.

I suppose this is one of those that is a good candidate for the _Won't Fix_ resolution. I will leave it to you guys to close it because I'm not sure if you want to add any special labels.

**Comment by ikonst** [comment]
First of all, probably correct to close the issue but I'll bring up the ["Type narrowing" docs](https://mypy.readthedocs.io/en/stable/type_narrowing.html) again. Is there a way we could better explain that we narrow only 'unconditionally' and recommend patterns for eliminating "entanglement"?

> Tracking these sort of cross-variable conditions in a type checker would add significant complexity and performance overhead and would be computationally infeasible in all but the most basic cases.

I had a similar thought: the complexity can grow exponentially. And if you were to contain it, the limit would be arbitrary and hard to reason about. A bit of a tangent, but this [Swift **compilation** error](https://www.cocoawithlove.com/blog/2016/07/12/type-checker-issues.html) comes up to mind:
```
error: expression was too complex to be solved in reasonable time; consider breaking up the expression into distinct sub-expressions
```

But I admit I'm curious whether there's a variant that _could_ somehow work, in whatever limited fashion, since I've "eliminated entanglement" many times, and the result would look like I'm placating a type-checker and get weird glances at code review time.
